File: prototypes/Threadpool.py
import threading


class ThreadPool:

    def __init__(self, max_workers):
        self.max_workers = max_workers
        self.tasks = []
        self.lock = threading.Lock()
        self.condition = threading.Condition(self.lock)
        self.threads = []
        self.stop_tasks = False
        self.kill = False

        for _ in range(max_workers):
            thread = threading.Thread(target=self.worker_loop)
            thread.start()
            self.threads.append(thread)
    
    def worker_loop(self):
        while True:
            task = None
            with self.condition:
                while len(self.tasks) == 0 and not self.stop_tasks:
                    self.condition.wait()
                if self.stop_tasks and len(self.tasks) == 0:
                    break
                if self.kill:
                    break
                task = self.tasks.pop(0)
                
            if task is not None:
                try:
                    task()
                except Exception as e:
                    print(f"Task failed: {e}")


    def close(self):
        print("Closing thread pool on job completion.")
        with self.condition:
            self.stop_tasks = True
            self.condition.notify_all()
        for thread in self.threads:
            thread.join()

    def submit(self, task):
        with self.condition:
            self.tasks.append(task)
            self.condition.notify()

    def kill(self):
        with self.condition:
            self.kill = True
            self.condition.notify_all()
        for thread in self.threads:
            thread.join()
        self.tasks.clear()
                
if __name__ == "__main__":
    import time
    pool = ThreadPool(6)
    for i in range(10):
        pool.submit(lambda i=i: (print(f"Task {i} started."), time.sleep(i), print(f"Task {i} finished.")))
    time.sleep(4)
    pool.submit(lambda: (print("Task 11 started."), [i for i in range(10000000)], print("Task 11 finished.")) )
    pool.close()
    print("All tasks completed.")

File: src/ThreadPool.cpp
#include <thread>
#include <vector>
#include <mutex>
#include <condition_variable>
#include <functional>
#include <queue>
#include <iostream>
#include "ThreadPool.h"


ThreadPool::ThreadPool(int max_workers)
{
    max_workers = max_workers;
    for(int w=0; w<max_workers; w++)
    {
        // using emplace_back rather than push back to skip creating a
        // thread using std::thread
        // this is used in capture class  as workerLoop is class member..
        workers.emplace_back([this](){ workerLoop();});
    }
    std::cout << "Thread pool created with " << max_workers << " workers" << std::endl;
}

ThreadPool::~ThreadPool()
{
    ThreadPool::stop_all();
}

void ThreadPool::workerLoop()
{   
    std::function<void()> task;
    while(true)
    {
        std::unique_lock<std::mutex> lock(mut);
        cv.wait(lock, [this]{
            return stop || !tasks.empty();
        });
        if(stop && tasks.empty())
        {
            break;
        }

        task = move(tasks.front());
        tasks.pop();
        lock.unlock();
        try
        {
            task();
        }
        catch(...)
        {
            std::cout << "Failure to execute tasks" << std::endl;
        }
        
    }
}

bool ThreadPool::add_task(std::function<void()> task)
{
    try
    {
        std::unique_lock<std::mutex> lock(mut);
        if(!stop)
        {
            tasks.emplace(task);
            cv.notify_one();
        }
        else 
        {
            std::cout << "Thread pool is stopped. Cannot add new tasks" << std::endl;
            return false;
        }
        
        lock.unlock();
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
        return false;
    }
    
    return true;
    
}

bool ThreadPool::stop_all()
{
    try
    {
        std::unique_lock<std::mutex> lock(mut);
        stop = true;
        cv.notify_all();
        lock.unlock();
        std::cout << "Notified all threads for stoppping"  << std::endl;
        for(auto &worker: workers)
        {
            worker.join();
        }

        return true;
    }
    
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
        return false;
    }
    
}



